File: dsp25.c
#include <graph.h>
#include <conio.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define DIM     128

float   fXR[DIM + 1], fXI[DIM + 1], fVR[DIM + 1], fVI[DIM + 1],
        fWR[DIM + 1], fWI[DIM + 1], fX[DIM + 1], fH[DIM + 1],
        fY[DIM + 1];

/*
*** WHEN COMPARING WITH THE TEXT, NOTE THAT IN THE C PROGRAMS **
******* FLOATING-POINT VARIABLES ARE PREFIXED WITH AN f, *******
************** INTEGER VARIABLES WITH AN i *********************
*/

void main (void)
    {
    /*
     **** PROGRAM NO. 25    FAST CONVOLUTION *******************
     */
    int    iCount, iCount2, iCount3, iXX, iYY, iM = 7,
           iP, iT, iLim1, iLim2, iLim3, iJ, iL;
    float  fMax, fXHMAG, fMag, fXT, fB1, fB2, fC1, fC2, fD,
           fPI = 3.141593, fX1, fX2, fARG, fCos1, fSin1;

    _clearscreen  (_GCLEARSCREEN);
    printf ("\n\nPROGRAM 25: Fast convolution\n\n");
    printf ("Press any key to start...\n\n");
    getch();

    /*
     *** DEFINE AND PLOT INPUT SIGNAL AND IMPULSE RESPONSE *****
     */
    for (iCount = 1; iCount <= 64; iCount++)
        fX[iCount] = 1.0;

    fH[1] = 1.0;
    fH[2] = 1.5;

    for (iCount = 3; iCount <= 64; iCount++)
        fH[iCount] = 1.5 * fH[iCount-1] - 0.85 * fH[iCount-2];

    _setvideomode( _MRES16COLOR );

    for (iCount = 1; iCount <= DIM; iCount++)
        {
        _moveto (iCount + 1, 70);
        _lineto (iCount + 1, (int)(70 - fX[iCount] * 25.0 + 0.5));
        }

    for (iCount = 1; iCount < DIM; iCount++)
        {
        _moveto (iCount + 190, 70);
        _lineto (iCount + 190, (int)(70 - fH[iCount] * 25.0 + 0.5));
        }

    /*
     *********** CONVOLVE IN TIME DOMAIN THEN PLOT *************
     */
    for (iCount = 1; iCount <= DIM; iCount++)
        for (iCount2 = 1; iCount2 < iCount; iCount2++)
            fY[iCount] += fX[iCount2] * fH[iCount - iCount2 + 1];

    for (iCount = 1; iCount <= DIM; iCount++)
        {
        _moveto (iCount + 100, 160);
        _lineto (iCount + 100, (int)(160 - fY[iCount] * 8.0 + 0.5));
        }

    printf ("Press any key to start fast convolution");
    getch();

    _clearscreen  (_GCLEARSCREEN);
    _setvideomode( _MRES16COLOR );

    for (iP = 1; iP <= 3; iP++)
        {
        if (iP == 1)
            {
            iT = 1;
            fD = 1.0;
            for (iCount = 1; iCount <= DIM; iCount++)
                {
                fXR[iCount] = fX[iCount];
                fXI[iCount] = 0.0;
                }
            }

        if (iP == 2)
            {
            iT = 1;
            fD = 1.0;
            for (iCount = 1; iCount <= DIM; iCount++)
                {
                fXR[iCount] = fH[iCount];
                fXI[iCount] = 0.0;
                }
            }

        if (iP == 3)
            {
            iT = -1;
            fD = (float) DIM;
            for (iCount = 1; iCount <= DIM; iCount++)
                {
                fXR[iCount] = fVR[iCount] * fWR[iCount] -
                                fVI[iCount] * fWI[iCount];
                fXI[iCount] = fVI[iCount] * fWR[iCount] +
                                fVR[iCount] * fWI[iCount];
                fXHMAG = pow ((pow (fVR[iCount], 2) +
                               pow (fVI[iCount], 2)) *
                               (pow (fWR[iCount], 2) +
                               pow (fWI[iCount], 2)), 0.5);
                _moveto (iCount, 160);
                _lineto (iCount, (int)(160 - fXHMAG / 4.0 + 0.5));
                }
            }
        /*
         **************** SHUFFLE INPUT DATA *******************
         */
        iLim1 = DIM - 1;
        iLim2 = DIM / 2;
        iJ = 1;

        for (iCount = 1; iCount <= iLim1; iCount++)
            {
            if ((float) iCount <= ((float) iJ - 0.01))
                {
                fX1 = fXR[iJ];
                fX2 = fXI[iJ];
                fXR[iJ] = fXR[iCount];
                fXI[iJ] = fXI[iCount];
                fXR[iCount] = fX1;
                fXI[iCount] = fX2;
                }
            iL = iLim2;
            while (1)
                {
                if ((float)iL > ((float)(iJ) - 0.01))
                    break;
                iJ -= iL;
                iL /= 2;
                }
            iJ += iL;
            }

        /*
         **************** IN PLACE TRANSFORMATION **************
         */
        for (iCount = 1; iCount <= iM; iCount++)
            {
            iLim1 = (int) pow (2, iCount - 1);
            iLim2 = (int) pow (2, (iM - iCount));

            for (iCount2 = 1; iCount2 <= iLim2; iCount2++)
                {
                for (iCount3 = 1; iCount3 <= iLim1; iCount3++)
                    {
                    iLim3 = (iCount3 - 1) + (iCount2 - 1) * 2 *
                        iLim1 + 1;
                    fB1 = fXR[iLim3];
                    fB2 = fXI[iLim3];
                    fC1 = fXR[iLim3 + iLim1];
                    fC2 = fXI[iLim3 + iLim1];

                    fARG = 2.0 * fPI * (float) (iCount3 - 1) *
                           (float) iLim2 / (float) DIM;
                    fCos1 = cosl (fARG);
                    fSin1 = sinl (fARG);
                    fX1 = fC1 * fCos1 + fC2 * fSin1 *
                        (float) iT;
                    fX2 = (0.0 - fC1) * fSin1 * (float) iT +
                        fC2 * fCos1;
                    fXR[iLim3] = fB1 + fX1;
                    fXI[iLim3] = fB2 + fX2;
                    fXR[iLim3 + iLim1] = fB1 - fX1;
                    fXI[iLim3 + iLim1] = fB2 - fX2;
                    }
                }
            }

        for (iCount = 1; iCount <= DIM; iCount++)
            {
            fXR[iCount] /= fD;
            fXI[iCount] /= fD;
            }

        /*
         * SAVE TRANSFORMS OF INPUT SIGNAL & IMPULSE RESPONSE **
         */
        if (iP == 1)
            {
            for (iCount = 1; iCount <= DIM; iCount++)
                {
                fVR[iCount] = fXR[iCount];
                fVI[iCount] = fXI[iCount];
                }
            }

        if (iP == 2)
            {
            for (iCount = 1; iCount <= DIM; iCount++)
                {
                fWR[iCount] = fXR[iCount];
                fWI[iCount] = fXI[iCount];
                }
            }

        /*
         ************** FIND PEAK VALUE FOR PLOT ***************
         */
        fMax = 0.0;
        for (iCount = 1; iCount <= DIM; iCount++)
            {
            fMag = pow (pow (fXR[iCount], 2) +
                pow (fXI[iCount], 2), 0.5);
            if (fMag > fMax)
                fMax = fMag;
            }

        /*
         ***************** SET PLOT OFFSETS ********************
         */
        switch (iP)
            {
            case 1 :
                iXX = 0;
                iYY = 70;
                break;

            case 2 :
                iXX = 190;
                iYY = 70;
                break;

            case 3 :
                iXX = 190;
                iYY = 160;
                break;
            }

        for (iCount = 1; iCount <= DIM; iCount++)
            {
            if (iP == 3)
                fXT = fXR[iCount];
            else
                fXT = pow (pow (fXR[iCount], 2) +
                    pow (fXI[iCount], 2), 0.5);

            _moveto (iCount + iXX, iYY);
            _lineto (iCount + iXX, (int)(iYY -
                fXT * 40.0 / fMax + 0.5));
            }
        }

    getch();
    _setvideomode( _DEFAULTMODE );
    
    /*
     ***********************************************************
     */
    }

File: dsp25.pas
(**********  PROGRAM NO.25       FAST CONVOLUTION  *************)
program DSP25;
uses graph3;
var X: array[1..128] of real;
var H: array[1..128] of real;
var Y: array[1..128] of real;
var XR: array[1..128] of real;
var XI: array[1..128] of real;
var VR: array[1..128] of real;
var VI: array[1..128] of real;
var WI: array[1..128] of real;
var WR: array[1..128] of real;
var A,K,N,M,L,P,T,D,LIM1,LIM2,J,I,R,LIM3,XX,YY: integer;
var XHMAG,X1,X2,PI,B1,B2,C1,C2,ARG,COS1,SIN1,MAX,MAG,XT: real;
label 1,2,3,4;
begin
    for K:=1 to 128 do begin X[K]:=0;H[K]:=0;Y[K]:=0;end;
    N:=128;M:=7;
(*****  DEFINE AND PLOT INPUT SIGNAL AND IMPULSE RESPONSE  *****)
    for K:=1 to 64 do begin X[K]:=1;end;
    H[1]:=1;H[2]:=1.5;
    for K:=3 to 64 do begin H[K]:=1.5*H[K-1]-0.85*H[K-2];end;
 graphmode;
    for K:=1 to N do begin
      draw(K,70,K,round(70-25*X[K]),1);end;
    for K:=1 to N do begin
      draw(K+190,70,K+190,round(70-25*H[K]),1);end;
(*************  CONVOLVE IN TIME DOMAIN, THEN PLOT  ************)
    for K:=1 to 128 do begin
      for L:=1 to K do begin
        Y[K]:=Y[K]+X[L]*H[K-L+1];
    end;end;
      for K:=1 to N do begin
        draw(K+100,160,K+100,round(160-8*Y[K]),1);end;
    writeln('ENTER ANY INTEGER TO START FAST CONVOLN');
    readln(A);
 clearscreen;
  for P:=1 to 3 do begin
      if P=1 then begin T:=1;D:=1;
        for K:=1 to N do begin
          XR[K]:=X[K];XI[K]:=0;end end;
      if P=2 then begin T:=1; D:=1;
        for K:=1 to N do begin
          XR[K]:=H[K];XI[K]:=0;end end;
      if P<3 then goto 1;
        T:=-1;D:=N;
        for K:=1 to N do begin
          XR[K]:=VR[K]*WR[K]-VI[K]*WI[K];
          XI[K]:=VI[K]*WR[K]+VR[K]*WI[K];
          XHMAG:=sqrt((VR[K]*VR[K]+VI[K]*VI[K])*
                                      (WR[K]*WR[K]+WI[K]*WI[K]));
          draw(K,160,K,round(160-XHMAG/4),1);
        end;
(*********************  SHUFFLE INPUT DATA  ********************)
 1: LIM1:=N-1;LIM2:=trunc(N/2);J:=1;
    for I:=1 to LIM1 do begin
      if I>J-1 then GOTO 2;
        X1:=XR[J];X2:=XI[J];
        XR[J]:=XR[I];XI[J]:=XI[I];XR[I]:=X1;XI[I]:=X2;
 2: L:=LIM2;
 3: if L>J-1 then goto 4;
      J:=J-L;L:=trunc(L/2);goto 3;
 4: J:=J+L;end;
(******************  IN-PLACE TRANSFORMATION  ******************)
    PI:=3.141593;
    for I:=1 to M do begin
      LIM1:=round(exp((I-1)*ln(2)));
      LIM2:=round(exp((M-I)*ln(2)));
      for L:=1 to LIM2 do begin
        for R:=1 to LIM1 do begin
          LIM3:=(R-1)+(L-1)*2*LIM1+1;
          B1:=XR[LIM3];B2:=XI[LIM3];
          C1:=XR[LIM3+LIM1];C2:=XI[LIM3+LIM1];
          ARG:=2*PI*(R-1)*LIM2/N;COS1:=cos(ARG);SIN1:=sin(ARG);
          X1:=C1*COS1+C2*SIN1*T;X2:=-C1*SIN1*T+C2*COS1;
          XR[LIM3]:=B1+X1;XI[LIM3]:=B2+X2;
          XR[LIM3+LIM1]:=B1-X1;XI[LIM3+LIM1]:=B2-X2;
      end;end;end;
    for K:=1 to N do begin
      XR[K]:=XR[K]/D;XI[K]:=XI[K]/D;end;
(***  SAVE TRANSFORMS OF INPUT SIGNAL AND IMPULSE RESPONSE  ****)
    if P=1 then begin
      for K:=1 to N do begin
        VR[K]:=XR[K];VI[K]:=XI[K];
    end end;
    if P=2 then begin
      for K:=1 to N do begin
        WR[K]:=XR[K];WI[K]:=XI[K];
    end end;
(****************  FIND PEAK VALUE FOR PLOT  *******************)
    MAX:=0;
    for K:=1 to N do begin
      MAG:=sqrt(XR[K]*XR[K]+XI[K]*XI[K]);
      if MAG>MAX then MAX:=MAG;
    end;
(**********************  SET PLOT OFFSETS  *********************)
    if P=1 then begin XX:=0;YY:=70 end;
    if P=2 then begin XX:=190;YY:=70 end;
    if P=3 then begin XX:=190;YY:=160 end;
(********  PLOT MAGNITUDE OF FFT, OR REAL PART OF IFFT  ********)
    for K:=1 to N do begin
        XT:=sqrt(XR[K]*XR[K]+XI[K]*XI[K]);
        if P=3 then XT:=XR[K];
        draw(K+XX,YY,K+XX,round(YY-XT*40/MAX),1);
      end;
  end;
end.
(***************************************************************)